packages/zoe/src/contractFacet/allocationMath.js

```diff
@@ -0,0 +1,71 @@
+// @ts-check
+import { AmountMath } from '@agoric/ertp';
+
+/**
+ * @callback Operation
+ *
+ * An operation such as add or subtract.
+ *
+ * @param {Amount | undefined} amount
+ * @param {Amount | undefined} diff
+ * @returns {Amount}
+ */
+
+/**
+ * Add amounts to an allocation, or subtract amounts from an
+ * allocation, by keyword. The operationFn performs the actual adding
+ * or subtracting.
+ *
+ * @param {Allocation} allocation
+ * @param {AmountKeywordRecord} amountKeywordRecord
+ * @param {Operation} operationFn
+ * @returns {AmountKeywordRecord}
+ */
+const doOperation = (allocation, amountKeywordRecord, operationFn) => {
+  const allKeywords = Object.keys({ ...allocation, ...amountKeywordRecord });
+  return Object.fromEntries(
+    allKeywords.map(keyword => [
+      keyword,
+      operationFn(allocation[keyword], amountKeywordRecord[keyword]),
+    ]),
+  );
+};
+
+/** @type {Operation} */
+const add = (amount, amountToAdd) => {
+  // Add if both are defined.
+  if (amount && amountToAdd) {
+    return AmountMath.add(amount, amountToAdd);
+  }
+  // If one is undefined (at least one will be defined), return the
+  // other, defined amount
+  return /** @type {Amount} */ (amount || amountToAdd);
+};
+
+/** @type {Operation} */
+const subtract = (amount, amountToSubtract) => {
+  if (amountToSubtract !== undefined) {
+    // @ts-ignore If amount is undefined, this will error appropriately.
+    return AmountMath.subtract(amount, amountToSubtract);
```

Comment:
I think this is a review of old code. The code currently is:

```
  if (amountToSubtract !== undefined) {
    assert(
      amount !== undefined,
      X`The amount could not be subtracted from the allocation because the allocation did not have an amount under the keyword ${keyword}.`,
    );
    assert(
      AmountMath.isGTE(amount, amountToSubtract),
      X`The amount to be subtracted ${amountToSubtract} was greater than the allocation's amount ${amount} for the keyword ${keyword}`,
    );
    return AmountMath.subtract(amount, amountToSubtract);
  }
```